src/macaron/repo_finder/provenance_finder.py

```diff
@@ -35,191 +37,350 @@ def __init__(self) -> None:
-                )
-                logger.error(msg)
-                raise ProvenanceAvailableException(msg)
-
-        provenance_filepaths = []
-        try:
-            with tempfile.TemporaryDirectory() as temp_dir:
-                for provenance_asset in provenance_assets:
-                    provenance_filepath = os.path.join(temp_dir, provenance_asset.name)
-                    if not provenance_asset.download(provenance_filepath):
-                        logger.debug(
-                            "Could not download the provenance %s. Skip verifying...",
-                            provenance_asset.name,
-                        )
-                        continue
-                    provenance_filepaths.append(provenance_filepath)
-        except OSError as error:
-            logger.error("Error while storing provenance in the temporary directory: %s", error)
-
-        provenances = []
-        witness_verifier_config = load_witness_verifier_config()
-
-        for provenance_filepath in provenance_filepaths:
             try:
-                provenance_payload = load_provenance_payload(provenance_filepath)
+                # Load the provenance file (provenance attestation).
+                provenance_payload = load_provenance_payload(download_path)
             except LoadIntotoAttestationError as error:
-                logger.error("Error while loading provenance: %s", error)
-                continue
+                logger.error("Error while loading provenance attestation: %s", error)
+                return []
 
-            if not is_witness_provenance_payload(provenance_payload, witness_verifier_config.predicate_types):
-                continue
+            signed_download_path = f"{download_path}.signed"
+            try:
+                # Load the other npm provenance file (publish attestation).
+                publish_payload = load_provenance_payload(signed_download_path)
+            except LoadIntotoAttestationError as error:
+                logger.error("Error while loading publish attestation: %s", error)
+                return [provenance_payload]
+
+            return [provenance_payload, publish_payload]
+
+    except OSError as error:
+        logger.error("Error while storing provenance in the temporary directory: %s", error)
+        return []
+
+
+def verify_npm_provenance(purl: PackageURL, provenance: list[InTotoPayload]) -> bool:
+    """Compare the unsigned payload subject digest with the signed payload digest, if available.
+
+    Parameters
+    ----------
+    purl: PackageURL
+        The PURL of the analysis target.
+    provenance: list[InTotoPayload]
+        The provenances to verify.
+
+    Returns
+    -------
+    bool
+        True if the provenance was verified, or False otherwise.
+    """
+    if len(provenance) != 2:
+        logger.debug("Expected unsigned and signed provenance.")
+        return False
```

Comment:
Am I understanding this correctly: this function can only work on `provenance: list[InTotoPayload` generated from `find_npm_provenance` (due to specific semantic requirements of that list).

Reply:
Yes. It is an example of a paired discovery and verification function.